utils/phaseUtils.py

```diff
@@ -0,0 +1,110 @@
+    The last element idx2Use contains a list of nonsequential bad phases
+
+    badPhaseCounts - container whose elements indicate bad GBT phase positions 
+    """
+
+    badPhasesAreSequential=False
+    minSeqPhase = -1
+    maxSeqPhase = -1
+    nBadPhases = len(badPhaseCounts)
+
+    idx2Use = [ x for x in range(0,nBadPhases) ]
+    
+    for idx1 in range(0,nBadPhases):
+        for idx2 in range(0,nBadPhases):
+            if not (idx1 > idx2):
+                continue
+            if( int(abs(badPhaseCounts[idx1]-badPhaseCounts[idx2])) == 1): #bad phases are sequential
+                minSeqPhase = int(min(badPhaseCounts[idx1],badPhaseCounts[idx2]))
+                maxSeqPhase = int(max(badPhaseCounts[idx1],badPhaseCounts[idx2]))
+                idx2Use.remove(idx1)
+                idx2Use.remove(idx2)
+                badPhasesAreSequential=True
+                break # exit inner loop
+            pass
+        if (badPhasesAreSequential):
+            break # exit outer loop
+        pass
+
+    return (badPhasesAreSequential, minSeqPhase, maxSeqPhase, idx2Use)
+
+def getPhaseFromLongestGoodWindow(badPhase, phaseCounts):
+    """
+    If there is only one bad phase, given by badPhase, in a phase scan this will
+    look for the longest good window and return a good phase value from that
+
+    badPhase    - the phase in the phase scan that is bad
+    phaseCounts - numpy array holding results of a phase scan
+    """
+    # only one bad phase point, search forward and backwards for most "good" phases
+    # wraparound needs to be handled by a circular range function
+    # phase == 15 needs to be handled, currently included in the sum
+    from gempython.tools.hw_constants import GBT_PHASE_RANGE 
+    WINDOW = 4
+    frange  = crange(int(badPhase+1),
+                     int(badPhase)+1+WINDOW,
+                     GBT_PHASE_RANGE)
+    brange  = crange(int(badPhase)-WINDOW,
+                     int(badPhase),
+                     GBT_PHASE_RANGE)
+    fsum = sum(phaseCounts.take(frange, mode='wrap')) # forward  sum
+    bsum = sum(phaseCounts.take(brange, mode='wrap')) # backward sum
+    if fsum > bsum:
+        return int((badPhase+WINDOW)%GBT_PHASE_RANGE)
+    elif bsum > fsum:
+        return int((badPhase-WINDOW)%GBT_PHASE_RANGE)
+    else:
+        ## choose the phase that doesn't require passing 15?
+        return int((badPhase-WINDOW)%GBT_PHASE_RANGE)
+
+def phaseIsGood(vfatBoard,vfat,phase):
+    """
+    Writes GBT phase to value 'phase' for given vfat
+    Returns true if this phrase is good
+
+    vfatBoard - Instance of HwVFAT class
+    vfat      - Number indicating vfat position in range [0,23]
+    phase     - Phase value to write in range [0,15]
+    """
+    
+    # Get hardware info
+    cardName = vfatBoard.parentOH.parentAMC.name
+    ohN      = vfatBoard.parentOH.link
+    phase    = vfatBoard.parentOH.vfatGBTPhases[vfat]
```

Comment:
Is it by design that you overwrite the phase value passed in as an argument to the function?

Reply:
hah no.  Will need to fix that.